tfhe/src/integer/server_key/radix_parallel/shift.rs

```diff
@@ -268,10 +365,99 @@ impl ServerKey {
+        let total_num_bits = num_bits_in_block * ct.blocks.len() as u64;
+
+        if shift > total_num_bits {
+            *ct = self.create_trivial_zero_radix(ct.blocks.len());
+            return;
+        }
+
+        // rotate right as the blocks are from LSB to MSB
+        ct.blocks.rotate_right(rotations);
+        // Every block below 'rotations' should be discarded
+        for block in &mut ct.blocks[..rotations] {
+            *block = self.key.create_trivial(0);
+        }
+
+        if shift_within_block == 0 || rotations == ct.blocks.len() {
+            return;
+        }
+
+        // Since we require that carries are empty,
+        // we can use the bivariate bps to shift and propagate in parallel at the same time
+        // instead of first shifting then propagating
+        //
+        // The first block is done separately as it does not
+        // need to recuperate the shifted bits from its previous block,
+        // and also that way is does not need a special case for when rotations == 0
+        let create_blocks_using_bivariate_pbs = || {
+            let lut = self
+                .key
+                .generate_accumulator_bivariate(|previous_block, current_block| {
+                    let current_block = current_block << shift_within_block;
+                    let previous_block = previous_block << shift_within_block;
+
+                    let message_of_current_block =
+                        current_block % self.key.message_modulus.0 as u64;
+                    let carry_of_previous_block =
+                        previous_block / self.key.message_modulus.0 as u64;
+                    message_of_current_block + carry_of_previous_block
+                });
+            let partial_blocks = ct.blocks[rotations..]
+                .par_windows(2)
+                .map(|blocks| {
+                    let (previous_block, current_block) = (&blocks[0], &blocks[1]);
+                    self.key.unchecked_apply_lookup_table_bivariate(
+                        previous_block,
+                        current_block,
+                        &lut,
+                    )
+                })
+                .collect::<Vec<_>>();
+            partial_blocks
+        };
+
+        let shift_last_block = || {
+            let mut block = ct.blocks[rotations].clone();
+            self.key
+                .scalar_left_shift_assign(&mut block, shift_within_block as u8);
+            block
+        };
+
+        let (partial_blocks, block) =
+            rayon::join(create_blocks_using_bivariate_pbs, shift_last_block);
+
+        // We now have to put the blocks back togeter
+        debug_assert_eq!(partial_blocks.len(), ct.blocks.len() - rotations - 1);
+        ct.blocks[rotations] = block;
+        for (block, shifted_block) in ct.blocks[rotations + 1..]
+            .iter_mut()
+            .zip(partial_blocks.into_iter())
+        {
+            *block = shifted_block
+        }
+
```

Comment:
could be parallelized potentially, not sure we would win much

Reply:
Entering a thread/switching context to a thread just to replace (replace + implicit drop) a value won't provide any

Reply:
right 🙄